datadogreceiver/translator.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datadogreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/datadogreceiver"
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"mime"
+	"net/http"
+	"strings"
+
+	"runtime/debug"
+
+	datadogpb "github.com/DataDog/datadog-agent/pkg/trace/exportable/pb"
+	"github.com/tinylib/msgp/msgp"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	semconv "go.opentelemetry.io/collector/semconv/v1.6.1"
+)
+
+func addResourceData(req *http.Request, rs *pcommon.Resource) {
+	version := ""
+	bi, ok := debug.ReadBuildInfo()
+	if ok {
+		version = bi.Main.Version
+	}
+	attrs := rs.Attributes()
+	attrs.Clear()
+	attrs.EnsureCapacity(3)
+	attrs.UpsertString("telemetry.sdk.name", "Datadog")
+	attrs.UpsertString("telemetry.sdk.version", version+"/"+"Datadog-"+req.Header.Get("Datadog-Meta-Tracer-Version"))
+	attrs.UpsertString("telemetry.sdk.language", req.Header.Get("Datadog-Meta-Lang"))
+}
+func toTraces(traces datadogpb.Traces, req *http.Request) ptrace.Traces {
+
+	dest := ptrace.NewTraces()
+	resSpans := dest.ResourceSpans()
+	rspan := resSpans.AppendEmpty()
+	resource := rspan.Resource()
+	addResourceData(req, &resource)
+
+	ils := rspan.ScopeSpans().AppendEmpty()
+
+	ils.Scope().SetName("Datadog-" + req.Header.Get("Datadog-Meta-Lang"))
+	ils.Scope().SetVersion(req.Header.Get("Datadog-Meta-Tracer-Version"))
+
+	for _, trace := range traces {
+
+		spans := ptrace.NewSpanSlice()
+		spans.EnsureCapacity(len(trace))
+		for _, span := range trace {
+			newSpan := spans.AppendEmpty()
+			newSpan.SetTraceID(uInt64ToTraceID(0, span.TraceID))
+			newSpan.SetSpanID(uInt64ToSpanID(span.SpanID))
+			newSpan.SetStartTimestamp(pcommon.Timestamp(span.Start))
+			newSpan.SetEndTimestamp(pcommon.Timestamp(span.Start + span.Duration))
+			newSpan.SetParentSpanID(uInt64ToSpanID(span.ParentID))
+			newSpan.SetName(span.Resource)
```

Comment:
Should this be span.Name?

Reply:
Possibly. Although personally the names output by span.Resource are much more interesting than span.Name. We probably should have both, but I think span.Name should be "Resource"